### ISSUE TITLE
Multiple assignment issue when a local used to index the left hand side is changed in the right hand side

### ISSUE THREAD

**Issue by pygy** [yuin/gopher-lua#158]
- [x] GopherLua is a Lua5.1 implementation. You should be familiar with Lua programming language. Have you read [Lua 5.1 reference manual](http://www.lua.org/manual/5.1/) carefully?
- [x] GopherLua is a Lua5.1 implementation.  In Lua, to keep it simple, **it is more important to remove functionalities rather than to add functionalities unlike other languages** . If you are going to introduce some new cool functionalities into the GopherLua code base and the functionalities can be implemented by existing APIs, It should be implemented as a library. 

Please answer the following before submitting your issue:

1. What version of GopherLua are you using? : 0.1
2. What version of Go are you using? : 1.9.2
3. What operating system and processor architecture are you using? : macOS 10.12.6, x86 64

Hi @yuin 

For multiple assignment, the [manual says](https://www.lua.org/manual/5.1/manual.html#2.4.3):

>  The assignment statement first evaluates all its expressions and only then are the assignments performed. Thus the code
>
>     i = 3
>     i, a[i] = i+1, 20
>
> sets `a[3]` to `20`, without affecting `a[4]` because the `i` in `a[i]` is evaluated (to `3`) before it is assigned `4`. 

while that exact sample works as described in gopher-lua, the following doesn't:

```
local i = 1
a = {}
a[i], i  = 20, i + 1

print(a[1] == 20, a[2] == nil, i == 2) -- false, false, true
```

`a[1]` is `nil`, and `a[2]` is `20` instead.

So the evaluation order is important, and it only occurs when `i` is a local.

This works fine:

```
local i = 1
a = {}
i, a[i] = i + 1, 20

print(a[1] == 20, a[2] == nil, i == 2) -- true, true, true
```

so does this:

```
i = 1
a = {}
a[i], i  = 20, i + 1

print(a[1] == 20, a[2] == nil, i == 2) -- true, true, true
```

**Comment by yuin** [comment]
This is an optimization bug. GopherLua evaluates all its expressions before performing assignment, but if a table index has already been loaded in a certain register, GopherLua uses this register and does not generate a new bytecode. If this table index is updated in same statement, the value in the register will also be updated. 

This issue has already been resolved in e37e1695a7de72e7d4c5559206a7fb3a6d5a5e88

**Comment by pygy** [comment]
Great :-)
